adafruit_ads1x15/ads1x15.py

```diff
@@ -0,0 +1,183 @@
+====================================================
+
+CircuitPython base class driver for ADS1015/1115 ADCs.
+
+* Author(s): Carter Nelson
+"""
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_ADS1x15.git"
+
+import time
+from micropython import const
+from adafruit_bus_device.i2c_device import I2CDevice
+
+# pylint: disable=bad-whitespace
+_ADS1X15_DEFAULT_ADDRESS            = const(0x48)
+_ADS1X15_POINTER_CONVERSION         = const(0x00)
+_ADS1X15_POINTER_CONFIG             = const(0x01)
+_ADS1X15_CONFIG_OS_SINGLE           = const(0x8000)
+_ADS1X15_CONFIG_MUX_OFFSET          = const(12)
+_ADS1X15_CONFIG_MODE_CONTINUOUS     = const(0x0000)
+_ADS1X15_CONFIG_MODE_SINGLE         = const(0x0100)
+_ADS1X15_CONFIG_COMP_QUE_DISABLE    = const(0x0003)
+_ADS1X15_CONFIG_GAIN = {
+    2/3: 0x0000,
+    1:   0x0200,
+    2:   0x0400,
+    4:   0x0600,
+    8:   0x0800,
+    16:  0x0A00
+}
+# pylint: enable=bad-whitespace
+
+class ADS1x15(object):
+    """Base functionality for ADS1x15 analog to digital converters."""
+
+    def __init__(self, i2c, address=_ADS1X15_DEFAULT_ADDRESS):
+        self.buf = bytearray(3)
+        self._gain = 1
+        self._data_rate = self._data_rate_default()
+        self._mode = _ADS1X15_CONFIG_MODE_SINGLE
+        self.i2c_device = I2CDevice(i2c, address)
+
+    @property
+    def data_rate(self):
+        """The data rate for ADC conversion in samples per second."""
+        return self._data_rate
+
+    @data_rate.setter
+    def data_rate(self, rate):
+        possible_rates = self.rates
+        if rate not in possible_rates:
+            raise ValueError("Data rate must be one of: {}".format(possible_rates))
+        self._data_rate = rate
+
+    @property
+    def rates(self):
+        """Possible data rate settings."""
+        raise NotImplementedError('Subclass must implement rates property.')
+
+    @property
+    def rate_config(self):
+        """Rate configuration masks."""
+        raise NotImplementedError('Subclass must implement rate_config property.')
+
+    @property
+    def gain(self):
+        """The ADC gain."""
+        return self._gain
+
+    @gain.setter
+    def gain(self, gain):
+        possible_gains = self.gains
+        if gain not in possible_gains:
+            raise ValueError("Gain must be one of: {}".format(possible_gains))
+        self._gain = gain
+
+    @property
+    def gains(self):
+        """Possible gain settings."""
+        g = list(_ADS1X15_CONFIG_GAIN.keys())
+        g.sort()
+        return g
+
+    @property
+    def mode(self):
+        """The ADC conversion mode."""
+        return self._mode
+
+    @mode.setter
+    def mode(self, mode):
+        if mode != _ADS1X15_CONFIG_MODE_CONTINUOUS or mode != _ADS1X15_CONFIG_MODE_SINGLE:
+            raise ValueError("Unsupported mode.")
+        self._mode = mode
+
+    def read(self, pin, is_differential=False):
+        """I2C Interface for ADS1x15-based ADCs reads.
```

Comment:
`_read()` was in the old version, and `read()` was added to mimic the MCP3xxx lib:
https://github.com/adafruit/Adafruit_CircuitPython_MCP3xxx/blob/master/adafruit_mcp3xxx/mcp3xxx.py#L78
and I also just mimic'd that doc string.

I feel like `read()` should be made more of a private func, since it's not really expected to be calling this directly. Users should use the `value` property on the channel. But I didn't want to deviate from the existing approach. So I generally agree, and more, but erred on the side of consistency.

Reply:
Yeah, I had the same thought about private vs public. But, that's a larger conversation outside of docstrings.

But, as the public function, this will generate in the RTD API which could be confusing. I'll resolve this, since the question is answered.

_(side note: in comparing to MCP32xxx, those API docs don't even build. will make an issue)_